### ISSUE TITLE
Support resource retry due to async nature of RT

### ISSUE THREAD

**Issue by KrystianMarek** [jfrog/terraform-provider-artifactory#30]
After moving resource from `/main.tf`:
```
terraform {
  required_providers {
    artifactory = {
      source  = "jfrog/artifactory"
      version = "~> 2.2.5"
    }
  }
}

# Configured via ENV
provider "artifactory" {}

resource "artifactory_local_repository" "npm-local" {
  key          = "npm-local"
  package_type = "npm"
}
```

to a module `/modules/core/main.tf`
```
terraform {
  required_providers {
    artifactory = {
      source  = "jfrog/artifactory"
      version = "~> 2.2.5"
    }
  }
}

resource "artifactory_local_repository" "npm-local" {
  key          = "npm-local"
  package_type = "npm"
}
```
so that the `/main.tf` looks like
```
terraform {
  required_providers {
    artifactory = {
      source  = "jfrog/artifactory"
      version = "~> 2.2.5"
    }
  }
}

# Configured via ENV
provider "artifactory" {}

module "core" {
  source = "./modules/core"
  providers = {
    artifactory = artifactory
  }
}
```

Terraform does not wait for Artifactory to delete a resource / repository:
```
[Pipeline] sh
16:53:24  + terraform apply -auto-approve
...
16:53:30  artifactory_local_repository.npm-local: Refreshing state... [id=npm-local]
...
16:53:35  artifactory_local_repository.npm-local: Destroying... [id=npm-local]
...
16:53:36  artifactory_local_repository.npm-local: Destruction complete after 1s
...
16:53:38  
16:53:38  Error: PUT https://SOME_SAAS_ARTIFACTORY.jfrog.io/artifactory/api/repositories/npm-local: 400 [{Status:400 Message:error when validating repository name: Case insensitive repository key already exists}]
16:53:38  
16:53:38    on modules/core/main.tf line 10, in resource "artifactory_local_repository" "npm-local":
16:53:38    10: resource "artifactory_local_repository" "npm-local" {
```

**Comment by chb0github** [comment]
@KrystianMarek - please supply your 
```terraform

# Configured via ENV
provider "artifactory" {}
```

I need a copy-and-paste example for reproducibility. I am going to run it against a local container that's brandnew.

**Comment by KrystianMarek** [comment]
@chb0github - I use ENV to configure the provider, like so - https://github.com/KrystianMarek/bug-terraform-artifactory-provider#test-environment

**Comment by peters95** [comment]
Hi @KrystianMarek this was never intended to work as a terraform module. We had internal discussions about whether it should be a provider or module with Hashicorp and determined that the provider route was appropriate. As such we never verified that this provider would work as a Terraform module. If you see things that need to be changed that are backward compatible with the provider I recommend you submit a PR with your desired changes and we can test and validate it on our side before we merge your desired changes. Thanks!

**Comment by neilbeukes** [comment]
Hi, We are running Artifactory and managing to it using Terraform also but on a larger scale. We are creating 50+ Repos  regularly and the Provider really gets confused on a regular basis. We believe it is to do with the async nature of its API (but this is just speculation). By running our `Terraform apply` with `-parallelism=1` It slows everything down greatly but it does stop these state drift's from happening. I would love if someone has a better solution as we are getting these `Repo already exists` or `Repo does not exist` errors far to regularly which then requires manual intervention. If need be I dont mind creating a new Issue.

**Comment by chb0github** [comment]
I am confused. Should this issue remain open or not?

**Comment by chb0github** [comment]
This will require a deeper dig - Do you have a clear example I can run to repeat the issue? If so, I can lean on the RT team for an answer. REST architecture is silent on parallelism and and there are 100+ reasons that an async implementation is either preferred or, forced upon the app (underlying data stores are frequently the drivers for such things as, depending on the choice, they aren't necessarily acid compliant for performance reasons). 

If might be best to implement (in the TF/go client) a BeB + retry. I will get back to you

**Comment by chb0github** [comment]
@KrystianMarek,

I looked at [your example](https://github.com/KrystianMarek/bug-terraform-artifactory-provider#test-environment) and it looks like you're describing another issue; a crash. If that's correct, please create another bug here

**Comment by chb0github** [comment]
@KrystianMarek - I am closing this issue as I have not heard back from you (and you have reached out to customer service anyways). When I looked at the explanation you created I noticed that you basically bundle up 2 problems in 1 - one is support for modules and the other was an index out of bounds panic (which I fixed in another PR). When creating issues, please just make 1-issue-1-ticket

As for the main-thrust of this ticket: What is needed is a more robust rest client like Resty that has built in support for retry and backoff. However, given the mixed nature and origin of this code base, that's a significant endeavour. I will pass this issue to our product people to make a decision on regarding resource allocation.

**Comment by chb0github** [comment]
BTW: Retry is available now in 2.3.6
